### ISSUE TITLE
warn on including package in both dependencies & devDependencies

### ISSUE THREAD

**Issue by smikes** [npm/npm#6725]
Spin off of https://github.com/npm/npm/issues/5403 which has a somewhat wider scope

> I discussed this a little with @isaacs; our consensus was that having the same packages in dependencies and devDependencies should be considered an error worth warning on (even in the absence of shrinkwrapping). It seems to me that what should happen is that when shrinkwrapping, npm should look to package.json only to see in which stanza a particular package belongs.

This may be relevant to development of multi-stage-install, so I am pulling it out as a separate issue /cc @iarna 

Feel free to close if this is already addressed in your plans; otherwise it should probably be tagged `feature-request`

**Comment by goloroden** [comment]
Would love to see such a warning, too :-)

**Comment by iarna** [comment]
The place to put this would be `lib/install/validate-tree.js` in `thenCheckTop`. You'd then compare `idealTree.package.dependencies` to `idealTree.package.devDependencies`.

**Comment by othiym23** [comment]
Another thing to keep in mind is that this warning is only relevant for the current top-level `package.json` – warnings shouldn't happen for any transitive dependencies, or people will get _way_ infuriated by being given non-actionable warnings. We'd love to see this added, though, because it's kind of surprising that npm doesn't already do this.

**Comment by TedYav** [comment]
On it 👍  Use NPM a lot and want to start contributing

**Comment by TedYav** [comment]
Okay, I found the function and wrote code to issue warning.
Some questions:

- Should I treat same package but different versions as equivalent?
- I know @iarna suggested putting this in `thenCheckTop` but I'm thinking it might do well in its own function. Right now I factored it out in a function as shown below, but I can also add it to its own function called `thenCheckForDuplicates` and add it to the chain at the top of the file.
- I gave it an arbitrary error code. Is there a central list of error codes or should I just give it a sensible one?

Thanks! Pardon me for asking so many questions :) I'll write tests up and submit a PR when I hear back.

Here's the function I wrote right now that works:

```
function checkForDuplicateDependencies(idealTree){
  for(let pkg in idealTree.package.dependencies){
    if(pkg in idealTree.package.devDependencies){
      var warnObj = new Error('The package ' + pkg + ' is included as both a dev and production dependency.')
      warnObj.code = 'EDUPLICATEDEP'
      idealTree.warnings.push(warnObj)
    }
  }
}
```
